Improve the HTTPS proxy example (#2973)
istio/istio.io: content/docs/examples/advanced-gateways/http-proxy/index.md

weight: 60
keywords: [traffic-management,egress]
---

> The [Configure an Egress Gateway](/docs/examples/advanced-gateways/egress-gateway/) example showed how you can direct

Author: Fixed in https://github.com/istio/istio.io/pull/2973/commits/2223452c4a53338feade62f918ac45d9eaa3b878

*   Setup Istio by following the instructions in the [Installation guide](/docs/setup/).

*   Start the [sleep]({{< github_tree >}}/samples/sleep) sample

>     which you will use as a test source for external calls via the proxy.

Author: Note that there is a guidance in istio.io to use "you" for the steps the reader takes. So it is not the example uses, it is the reader who **will** use the sleep sample after he/she starts it.

So which one is better?

1. Start the sleep sample which you use in this example
2. Start the sleep sample which you **will** use in this example

cc @frankbu

>     which you will use as a test source for external calls via the proxy.

Reviewer: Best is:

    To have test source for external calls via the proxy, start the sleep sample.

Before the reader starts the sample, they should have read a sentence similar to:

    This example uses the sleep sample as a test source for external calls via the proxy.

That way you let readers know how the sleep sample is used throughout and you don't need to mention it on each step.

>     which you will use as a test source for external calls via the proxy.

Author: Applied in https://github.com/istio/istio.io/pull/2973/commits/a359411b2acd1ecfd00b59f5a403fce037d25e8f

    $ kubectl apply -f <(istioctl kube-inject -f @samples/sleep/sleep.yaml@)
    {{< /text >}}

    Note that you can use as a test source any pod that has `curl` installed.

*   Create the `SOURCE_POD` environment variable to store the name of the source pod for sending requests to external
    services:

    {{< text bash >}}
    $ export SOURCE_POD=$(kubectl get pod -l app=sleep -o jsonpath={.items..metadata.name})
    {{< /text >}}

## Deploy an HTTPS proxy

For the sake of this example only, to simulate a legacy proxy, you deploy an HTTPS proxy inside your cluster.

> Also, to simulate a more realistic proxy that is running outside of your cluster, you will address the pod of the proxy

Reviewer: Avoid using words with little to no semantic value to simplify your text. Use the imperative present tense instead of the future tense. Lastly, I suggest using the possessive to move the noun "pod" closer to the pronoun "its" to make the reference more evident:
 ```suggestion
To simulate a more realistic proxy running outside of your cluster, address the proxy's pod
```

Author: `Also` is added to connect with the previous sentence:

> To simulate a legacy proxy... Moreover, [to be closer to the real life situation], you will address the proxy's pod by its IP address

Please advise.

Author: > Use the imperative present tense instead of the future tense

The imperative could confuse the readers - they can ask themselves: "What should I do to address the proxy's pod by its IP address?". 

The meaning here is "in the following instructions, when you will address the pod, you will address it by its IP address, as we will instruct you".

Please advise.

Author: Alternatively, the meaning is: in the following steps you will address the pod by its IP, do not be surprised - you will do it to simulate a real life situation better.

Reviewer: Let me address your concerns in order:
1. The use of `Also`, `Moreover`, and similar sentence connecting words:
As we learn to write, we are taught that the sentences in a paragraph must be linked using these words. What most of us forget, is that we are also taught to do so sparingly and that not all lead words are the same. Readers connect sentences in a paragraph with an implicit `AND`. My advice to you would be to only use leading words when the connection differs from `AND`. For example, to connect two sentences in a paragraph with `BUT` I would use "However, ..."
2. The use of imperative:
I agree fully, readers will ask  "How do I address the proxy's pod by its IP address?". If you know this, my advise would be to link to the instructions on how to do that. The imperative will help you identify those questions and then ask: Will my audience know the answer or not? If not, link to the answer, define it briefly or link to a definition. 
3. The future meaning conflicts with the use of present tense: 
I understand that from your perspective, you are describing the actions readers will take and, thus, feel the need to use the future tense in your text. However, I would advise you to switch from a descriptive approach, writing a descriptions of the future actions to take, to a prescriptive approach, writing the actions readers need to take at that moment. What action do readers need to take at the beginning of your document: Keep reading! Therefore, I suggest:
Keep reading to learn how to address the pod by its IP and how this approach allows you to simulate a real life situation better.

Hopefully, I've been clear enough and answered your questions. Let me know if further discussion is needed.

Author: Note that addressing the pod by IP is done as part of the following steps. The sentence means: when you will perform the steps below, note that you will address the pod by IP (as part of the steps, not for the sake of addressing the pod by IP).

    $ kubectl apply -f <(istioctl kube-inject -f @samples/sleep/sleep.yaml@)
    {{< /text >}}

    You can use any pod with `curl` installed as a test source.

*   To send requests to external services, create the `SOURCE_POD` environment variable to store the name of the source
    pod:

    {{< text bash >}}
    $ export SOURCE_POD=$(kubectl get pod -l app=sleep -o jsonpath={.items..metadata.name})
    {{< /text >}}

## Deploy an HTTPS proxy

To simulate a legacy proxy and only for this example, you deploy an HTTPS proxy inside your cluster.
Also, to simulate a more realistic proxy that is running outside of your cluster, you will address the proxy's pod
by its IP address and not by the domain name of a Kubernetes service.
This example uses [Squid](http://www.squid-cache.org) but you can use any HTTPS proxy that supports HTTP Connect.

1.  Create a namespace for the HTTPS proxy. Since the automatic sidecar injection is disabled in Istio by default,
    Istio sidecars will not be injected and Istio will not control traffic in this namespace. You need this behavior to

>     simulate the proxy outside of the cluster.

Author: Addressed in https://github.com/istio/istio.io/pull/2973/commits/29f2e1c8c1629fdc3d3e3b983c6e3b6c72273b04

## Understanding what happened

In this example you configured access to an external legacy HTTPS proxy from the Istio service mesh. To simulate a
remote HTTPS proxy, you deployed an HTTPS proxy in a separate namespace, without Istio automatic sidecar injection.
In addition, for simulation of a remote proxy, you addressed the HTTPS proxy by its IP address and not by the domain

> name of a Kubernetes service.

Reviewer: I love the fact that you are summarizing what the readers did on the example. I would advise you to do so as a numbered list instead:

```suggestion
In this example, you took the following steps to configure access to an external legacy HTTPS proxy from an Istio service mesh:

1. Deployed an HTTPS proxy in a separate namespace without Istio automatic sidecar injection to simulate a remote HTTPS proxy.
2. Addressed the HTTPS proxy by its IP address and not by the domain name of a Kubernetes service to maintain the simulation of a remote proxy.
```
You would need to continue with the content of the next paragraph. The numbered steps provide additional clarity to readers. If you can, try to be more high-level with this overarching flow. You already covered the nit and grit of a lot of it above.
